spec/abci++/v3.md

```diff
@@ -0,0 +1,201 @@
+
+```go
+upon ⟨PROPOSAL, h_p, round_p, v_header, −1) from proposer(h_p, round_p) while step_p = propose do {
+    prevote_nil ← false
+    // valid is Tendermints validation, ABCI.VerifyHeader is the applications
+    if valid(v_header) ∧ ABCI.VerifyHeader(h_p, v_header) ∧ (lockedRound_p = −1 ∨ lockedValue_p = id(v_header)) {
+        wait to receive proposal v corresponding to v_header
+        // We split up the app's header verification from the remainder of its processing of the proposal
+        if ABCI.ProcessProposal(h_p, v).accept {
+            broadcast ⟨PREVOTE, h_p, round_p, id(v)⟩ 
+        } else {
+            prevote_nil ← true
+            // Include any slashing evidence that may be sent in the process proposal response
+            for evidence in ABCI.ProcessProposal(h_p, v).evidence_list {
+                broadcast ⟨EVIDENCE, evidence⟩ 
+            }
+        }
+    } else {
+        prevote_nil ← true
+    }
+    if prevote_nil {
+        broadcast ⟨PREVOTE, h_p, round_p, nil⟩ 
+    }
+    step_p ← prevote
+}
+```
+
+In the case where the node is locked on a round, the following is ran:
+
+```go
+upon ⟨PROPOSAL, h_p, round_p, v_header, vr⟩
+  from proposer(h_p, round_p)
+  AND 2f + 1 ⟨PREVOTE, h_p, vr, id(v_header)⟩ 
+  while step_p = propose ∧ (vr ≥ 0 ∧ vr < round_p) do {
+    prevote_nil ← false
+    if valid(v) ∧ ABCI.VerifyHeader(h_p, v.header) ∧ (lockedRound_p ≤ vr ∨ lockedValue_p = v) {
+        wait to receive proposal v corresponding to v_header
+        // We split up the app's header verification from the remainder of its processing of the proposal
+        if ABCI.ProcessProposal(h_p, v).accept {
+            broadcast ⟨PREVOTE, h_p, round_p, id(v)⟩
+        } else {
+            prevote_nil ← true
+            // Include any slashing evidence that may be sent in the process proposal response
+            for evidence in ABCI.ProcessProposal(h_p, v).evidence_list {
+                broadcast ⟨EVIDENCE, evidence⟩ 
+            }
+        }
+    } else {
+        prevote_nil ← true
+    }
+    if prevote_nil {
+        broadcast ⟨PREVOTE, h_p, round_p, nil⟩
+    }
+    step_p ← prevote
+}
+```
+
+### Prevote timeout
+
+Upon receiving 2f + 1 prevotes, setup a timeout.
+
+```go
+upon 2f + 1 ⟨PREVOTE, h_p, vr, -1⟩ 
+  with step_p = prevote for the first time, do {
+    schedule OnTimeoutPrevote(h_p, round_p) to be executed after timeoutPrevote(round_p)
+}
+```
+
+with OnTimeoutPrevote defined as:
+
+```go
+function OnTimeoutPrevote(height, round) {
+    if (height = h_p && round = round_p && step_p = prevote) {
+        precommit_extension ← ABCI.ExtendVote(h_p, round_p, nil)
+        broadcast ⟨PRECOMMIT, h_p, round_p, nil, precommit_extension⟩
+        step_p ← precommit
+    }
+}
+```
+
+### Receiving enough prevotes to precommit
+
+The following code is ran upon receiving 2f + 1 prevotes for the same block
+
+```go
+upon ⟨PROPOSAL, h_p, round_p, v, *⟩
```

Comment:
PROPOSAL message contains full block proposal in this rule, while above it is only header. As mentioned above, we can make it clear by having in addition to PROPOSAL, also PROPOSAL_HEADER message and then have a separate rule for PROPOSAL_HEADER message.